### ISSUE TITLE
lower BLEU

### ISSUE THREAD

**Issue by SwordYork** [nyu-dl/dl4mt-c2c#1]
Thanks for your great work! I have downloaded the trained model and evaluated on csen model. However I got 22.70 and 23.58 on 'dev' and 'test1' respectively which is lower than the paper (about 0.6). I have no idea what causes the issue, do I need to train the model further?

**Comment by jaseleephd** [comment]
Hi @SwordYork, that is a bit weird. is it the bi-char2char model? what's the model number? is it 1505000?

**Comment by jaseleephd** [comment]
Hi @SwordYork, I just doubled-checked by decoding the trained model for CS-EN and computing the BLEU score. Two things:

1) I uploaded the wrong bi-char2char model for CS-EN. Apologies. The model number should be 1505000, not 785000 (which is the version you find when you download & untar the ball). With the 1505000 model, I confirm that I get 23.38 BLEU on dev and 24.08 BLEU on test1 (as reported in the paper). 

Please download the correct model [here](https://drive.google.com/open?id=0BxmEQ91VZAPQT3VxSGljbmFlNUU). Will shortly update the original tarball to include this one instead of the 785000 model.

2) I took the 785000 CS-EN bi-char2char model (the version that was included in the tarball) and computed BLEU scores for the development and the test1 set, and I'm getting different BLEU scores than you. I'm getting 22.85 BLEU for dev and 23.38 BLEU for test1. So I'm suspecting you might be using a different reference file when computing BLEU? You should be using **wmt15/csen/dev/newstest2013-src.en.tok** for dev and **wmt15/csen/test/newstest2014-csen-src.en.tok** for test1 respectively. (from the preprocessed wmt corpus that I uploaded in the repo)

Please let me know whether you managed to reproduce the results with the instructions above. 

Cheers
Jason

**Comment by SwordYork** [comment]
Thanks, @jasonleeinf . I'm using the same reference file, maybe different version of theano or cuda that causes the different BLEU. I also find that model for fi-en produces a lower BLEU, maybe the same problem.

I have implemented another fully character NMT, however fully RNN based. Could I compare the BLEU scores when both trained after 785000 updates? I think it is too costly to train for 1500000.

**Comment by jaseleephd** [comment]
The CS-EN results we report are obtained from that particular model (1505000), so you should compare your results with that number. We stopped training only when the validation BLEU score stopped improving, maybe you could do the same instead of fixing the number of updates for your model?

The CS-EN sentences are on average much shorter than the other three language pairs, hence the CS-EN model is much faster to train.

p.s. I'm not sure different theano/CUDA versions could give different BLEU scores from the same pre-trained model. I got consistent results when I decoded my models from different machines, so I'm not entirely sure why you're getting different BLEU scores. 

If you could send me your translation outputs for dev/test1 I can compare them with mine and try and see what the difference is?

**Comment by jaseleephd** [comment]
FI-EN as well? That's really weird... What's your CUDA/theano version?

**Comment by SwordYork** [comment]
I got BLEU = 13.40, 46.0/18.5/8.7/4.4 (BP=1.000, ratio=1.014, hyp_len=32952, ref_len=32494)  which is similar to the Figure 2 in the paper. I think such a big difference could not caused by version.  The model is trained for 260000 iterations.

**Comment by jaseleephd** [comment]
I'm guessing this is FI-EN? dev or test2? Did you train the bi-char2char model yourself or is this BLEU score from the pre-trained model?

**Comment by SwordYork** [comment]
Yes, it's FI-EN and tested on dev. I just test on your trained bi-char2char model. However, it matches the result in Figure 2.

**Comment by jaseleephd** [comment]
Figure 2 is plotted with exponential smoothing for better interpretability. You should get the same numbers from Table 5 (14.18 for dev, 13.10 for test2). I'll be more than happy to examine your translation output if you could send them to jasonlee@inf.ethz.ch

**Comment by SwordYork** [comment]
I have sent it to you. I notice that it generates several empty sentences, maybe it is the problem.

**Comment by jaseleephd** [comment]
Thanks! I'll get back to you by tomorrow

**Comment by jaseleephd** [comment]
Hi @SwordYork, I decoded FI-EN Dev from scratch using the pre-trained model (260000), and it gives me the reported BLEU score: **BLEU = 14.18, 47.0/19.2/9.3/4.8 (BP=1.000, ratio=1.005, hyp_len=32667, ref_len=32494)**. Please see my translation output [here](https://drive.google.com/open?id=0BxmEQ91VZAPQUV9sazlpMHh5VUU) .

I also compared my system output with [yours](https://drive.google.com/open?id=0BxmEQ91VZAPQblB1NE1uUW9ra0U) and noticed a couple of differences.

1. My system output and yours look completely different. Not just a few sentences, every sentence. So either we're using a different model to decode from (which I don't think is the case, as we're both using the same 260000 model) or we're using different source files.

2. Like you mentioned, your output contains several completely empty sentences. This never happens with any of my models. 

I have one question for you. When you run `translate.py`, did you provide an argument for `-source`? Did you give any additional arguments apart from that? Can you post here the exact command you ran to translate please?

Thanks!

**Comment by SwordYork** [comment]
I didn't pass an argument for source, just use `which`. I also found that the decode speed is much faster than the paper, is it related to the beamwidth? Here is my command:
`python2 translate/translate_char2char.py -model /home/york/Program/c2c/models/bi-char2char/fi_en/one-multiscale-conv-four-hw-gru-12345678-200-200-250-250-300-300-300-300-55-pool3-512-128-False.npz -translate fi_en -saveto fi_en_t`

The two translated files have similar content, so it can't be different source files.
I'm using Theano '0.8.2.dev-RELEASE' and CUDA 7.5, I will test on the last CUDA tomorrow.

**Comment by jaseleephd** [comment]
You should add `-which dev` and use `-model .../one-multiscale-conv-four-hw-gru-12345678-200-200-250-250-300-300-300-300-55-pool3-512-128-False.grads.260000.npz` instead of `....False.npz`. The models with `.grads` in their names are the best performing models. Apologies if this wasn't clear from the README!

**Comment by SwordYork** [comment]
Isn't `dev` the default? I have misunderstood that `grads` represents gradient. I will test it tomorrow. Thanks!

**Comment by jaseleephd** [comment]
Okay I just tried decoding FI-EN Dev using the non-optimal model `...False.npz` that you used, and got the same output as you: BLEU = 13.40, 46.0/18.5/8.7/4.4 (BP=1.000, ratio=1.014, hyp_len=32952, ref_len=32494). This is the wrong model, sorry about the confusion.

Will add a note in the README to use the `.grads` model. Thanks a lot for pointing this out! Will also update the CS-EN bi-char2char model in the tarball.

Please let me know if you can reproduce the results with the optimal model tomorrow! Thanks.

**Comment by SwordYork** [comment]
@jasonleeinf I have tested on the optimal model, and obtain the same result. Thanks!
`BLEU = 14.18, 47.0/19.2/9.3/4.8 (BP=1.000, ratio=1.005, hyp_len=32667, ref_len=32494)
`

**Comment by jaseleephd** [comment]
Thanks for letting me know!
